libraries/GCS_MAVLink/GCS_FTP.cpp

```diff
@@ -0,0 +1,600 @@
+                                      "FTP", 1024, AP_HAL::Scheduler::PRIORITY_IO, 0)) {
+        goto failed;
+    }
+
+    return true;
+
+failed:
+    delete ftp.requests;
+    ftp.requests = nullptr;
+    delete ftp.replies;
+    ftp.replies = nullptr;
+
+    return false;
+    
+}
+
+void GCS_MAVLINK::handle_file_transfer_protocol(const mavlink_message_t &msg) {
+    if(ftp_init()) {
+        mavlink_file_transfer_protocol_t packet;
+        mavlink_msg_file_transfer_protocol_decode(&msg, &packet);
+
+        struct pending_ftp request;
+
+        request.chan = chan;
+        request.seq_number = *(uint16_t *)packet.payload;
+
+        request.session = packet.payload[2];
+        request.opcode = static_cast<FTP_OP>(packet.payload[3]);
+        request.size = packet.payload[4];
+        request.req_opcode = static_cast<FTP_OP>(packet.payload[5]);
+        request.burst_complete = packet.payload[6];
+        request.offset = *(uint32_t *)(&packet.payload[8]);
+        request.sysid = msg.sysid;
+        request.compid = msg.compid;
+        memcpy(request.data, &packet.payload[12], sizeof(packet.payload) - 12);
+
+        if (!ftp.requests->push(request)) {
+            // dropping the message, no buffer space to queue it in
+            // we could NACK it, but that can lead to GCS confusion, so we're treating it like lost data
+        }
+    }
+}
+
+void GCS_MAVLINK::send_ftp_replies(void) {
+    if (ftp.replies != nullptr) {
+        for (uint8_t i = 0; i < 20; i++) {
+            struct pending_ftp reply;
+            uint8_t payload[251] = {};
+            if (ftp.replies->peek(reply) && (reply.chan == chan)) {
+                    if ((i > 0) && comm_get_txspace(chan) < (2 * (packet_overhead() + MAVLINK_MSG_ID_FILE_TRANSFER_PROTOCOL_LEN))) {
+                        // if this isn't the first packet we have to leave deadspace for the next message
+                        break;
+                    }
+                    if (!HAVE_PAYLOAD_SPACE(chan, FILE_TRANSFER_PROTOCOL)) {
+                        return;
+                    }
+                    ((uint16_t *)payload)[0] = reply.seq_number;
+                    payload[2] = reply.session;
+                    payload[3] = static_cast<uint8_t>(reply.opcode);
+                    payload[4] = reply.size;
+                    payload[5] = static_cast<uint8_t>(reply.req_opcode);
+                    payload[6] = reply.burst_complete ? 1 : 0;
+                    *(uint32_t *)(&payload[8]) = reply.offset;
+                    memcpy(&payload[12], reply.data, sizeof(reply.data));
+                    mavlink_msg_file_transfer_protocol_send(
+                        reply.chan,
+                        0, reply.sysid, reply.compid,
+                        payload);
+                    ftp.replies->pop(reply);
+            } else {
+                break;
+            }
+        }
+    }
+}
+
+void GCS_MAVLINK::ftp_error(struct pending_ftp &response, FTP_ERROR error) {
```

Comment:
can this be `const`?  Any other methods likewise?

Reply:
Nope, or at least not at the moment. It's a static method, which by definition can't be const. The only reason the helper is even a member of GCS_MAVLINK is because the pending_ftp struct is part of GCS_MAVLINK. It could be changed to be const, but technically you can cast away the constness internally anyways, whereas static doesn't have a pointer to work on for that.